solvers/ADMM.py

```diff
@@ -56,12 +52,23 @@ def run(self, callback):
                                                         prepend=0))
         while callback(u):
             z_old = z
-            u_tmp = (Aty + np.diff(mu, append=0, prepend=0)
-                     - gamma * np.diff(z, append=0, prepend=0))
-            if isinstance(self.A, np.ndarray):
-                u = np.ravel(AtA_gDtD_inv @ u_tmp)
+            if self.data_fit == 'quad':
+                u_tmp = (Aty + np.diff(mu, append=0, prepend=0)
+                         - gamma * np.diff(z, append=0, prepend=0))
+                if isinstance(self.A, np.ndarray):
+                    u = np.ravel(AtA_gDtD_inv @ u_tmp)
+                else:
+                    u, _ = cg(AtA_gDtD, u_tmp, x0=u, tol=tol_cg)
             else:
-                u, _ = cg(AtA_gDtD, u_tmp, x0=u, tol=tol_cg)
+                u_tmp = (np.diff(mu, append=0, prepend=0)
+                         - gamma * np.diff(z, append=0, prepend=0))
+
+                def func(u):
+                    return abs(self.grad(self.A, u)
+                               - gamma * np.diff(np.diff(u),
+                                                 append=0, prepend=0)
+                               - u_tmp).sum()
+                u = minimize(func, x0=u, method='L-BFGS-B', tol=tol_cg).x
```

Comment:
@EnLAI111 you should pass the gradient function to the minimize otherwise it approximates it with finite differences and it's very imprecise